File: lib/services/project_detail_preferences.dart
import 'package:shared_preferences/shared_preferences.dart';

/// Estado de UI do detalhe do projeto (seções recolhidas, concluídas expandidas).
class ProjectDetailPreferences {
  static String collapsedSectionsKey(String projectId) =>
      'proj_detail_collapsed_sections_$projectId';

  static String completedExpandedKey(String projectId) =>
      'proj_detail_completed_expanded_$projectId';

  static Future<Set<String>> collapsedSectionIds(String projectId) async {
    final prefs = await SharedPreferences.getInstance();
    return (prefs.getStringList(collapsedSectionsKey(projectId)) ?? []).toSet();
  }

  static Future<void> setCollapsedSectionIds(
    Set<String> ids,
    String projectId,
  ) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setStringList(collapsedSectionsKey(projectId), ids.toList());
  }

  static Future<bool> completedExpanded(
    String projectId, {
    bool defaultValue = false,
  }) async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getBool(completedExpandedKey(projectId)) ?? defaultValue;
  }

  static Future<void> setCompletedExpanded(bool value, String projectId) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setBool(completedExpandedKey(projectId), value);
  }
}

File: stacked-ios/Stacked/Services/ProjectDetailPreferences.swift
import Foundation

// Paridade lib/services/project_detail_preferences.dart — estado de UI do detalhe do projeto.
enum ProjectDetailPreferences {
  static func collapsedSectionsKey(projectId: String) -> String {
    "proj_detail_collapsed_sections_\(projectId)"
  }

  static func completedExpandedKey(projectId: String) -> String {
    "proj_detail_completed_expanded_\(projectId)"
  }

  static func collapsedSectionIds(projectId: String) -> Set<String> {
    let key = collapsedSectionsKey(projectId: projectId)
    guard let raw = UserDefaults.standard.array(forKey: key) as? [String] else { return [] }
    return Set(raw)
  }

  static func setCollapsedSectionIds(_ ids: Set<String>, projectId: String) {
    UserDefaults.standard.set(Array(ids), forKey: collapsedSectionsKey(projectId: projectId))
  }

  static func completedExpanded(projectId: String, default defaultValue: Bool = false) -> Bool {
    let key = completedExpandedKey(projectId: projectId)
    guard UserDefaults.standard.object(forKey: key) != nil else { return defaultValue }
    return UserDefaults.standard.bool(forKey: key)
  }

  static func setCompletedExpanded(_ value: Bool, projectId: String) {
    UserDefaults.standard.set(value, forKey: completedExpandedKey(projectId: projectId))
  }

  // MARK: - Subtarefas inline expandidas (global)

  private static let expandedSubtaskTasksKey = "subtask_expanded_task_ids"

  static func expandedSubtaskTaskIds() -> Set<String> {
    guard let raw = UserDefaults.standard.array(forKey: expandedSubtaskTasksKey) as? [String] else { return [] }
    return Set(raw)
  }

  static func isSubtaskListExpanded(taskId: String) -> Bool {
    expandedSubtaskTaskIds().contains(taskId)
  }

  static func setSubtaskListExpanded(_ expanded: Bool, taskId: String) {
    var ids = expandedSubtaskTaskIds()
    if expanded, !taskId.isEmpty {
      ids.insert(taskId)
    } else {
      ids.remove(taskId)
    }
    UserDefaults.standard.set(Array(ids), forKey: expandedSubtaskTasksKey)
  }
}
